### PR TITLE
Add table-driven tests to x/tokenfactory

### PR DESCRIPTION
**Pull request by vuong177**
Closes: #1418
## What is the purpose of the change

I add table-driven tests for x/tokenfactory. It includes `TestMintDenom`, `TestBurnDenom`, `TestChangeAdminDenom`, `TestCreateDenom`

### CODE REVIEW

File: x/tokenfactory/keeper/admins_test.go [hunk 1]
```diff
@@ -72,3 +74,221 @@ func (suite *KeeperTestSuite) TestAdminMsgs() {
...
+func (suite *KeeperTestSuite) TestMintDenom() {
+	suite.SetupTest()
+	var (
+		denom     string
+		msgServer types.MsgServer
+		addr0bal  = int64(0)
+	)
+
+	// Create a denom
+	msgServer = keeper.NewMsgServerImpl(*suite.App.TokenFactoryKeeper)
```
Comment (alexanderbez) [inline x/tokenfactory/keeper/admins_test.go:87]: 1. We should probably define and setup `msgServer` in the `suite` itself upon `SetupTest`
2. Also, why are we making explicit calls to `SetupTest`? Testify will automatically call `SetupTest` prior to each test.

Comment (vuong177) [reply x/tokenfactory/keeper/admins_test.go:87]: Sure. I'm going to change it

File: x/tokenfactory/keeper/keeper_test.go [hunk 2]
```diff
@@ -7,13 +7,16 @@ import (
...
+	"github.com/osmosis-labs/osmosis/v7/x/tokenfactory/keeper"
 	"github.com/osmosis-labs/osmosis/v7/x/tokenfactory/types"
 )
 
 type KeeperTestSuite struct {
 	apptesting.KeeperTestHelper
 
-	queryClient types.QueryClient
+	queryClient  types.QueryClient
+	msgServer    types.MsgServer
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/keeper_test.go:18]: oh interesting, we should definitely add this to our default test suite code structure everywhere!

File: x/tokenfactory/keeper/keeper_test.go [hunk 3]
```diff
@@ -7,13 +7,16 @@ import (
...
+	defaultDenom string
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/keeper_test.go:19]:
```suggestion
	// defaultDenom is on the suite, as it depends on the creator test address.
	defaultDenom string
```

File: x/tokenfactory/keeper/createdenom_test.go [hunk 4]
```diff
@@ -50,11 +48,70 @@ func (suite *KeeperTestSuite) TestMsgCreateDenom() {
...
 	// Make sure that an address with a "/" in it can't create denoms
-	res, err = msgServer.CreateDenom(sdk.WrapSDKContext(suite.Ctx), types.NewMsgCreateDenom("osmosis.eth/creator", "bitcoin"))
+	res, err = suite.msgServer.CreateDenom(sdk.WrapSDKContext(suite.Ctx), types.NewMsgCreateDenom("osmosis.eth/creator", "bitcoin"))
 	suite.Require().Error(err)
 }
+
+func (suite *KeeperTestSuite) TestCreateDenom() {
+	for _, tc := range []struct {
+		desc     string
+		malleate func()
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/createdenom_test.go:63]:
```suggestion
		setup func()
```

File: x/tokenfactory/keeper/createdenom_test.go [hunk 5]
```diff
@@ -50,11 +48,70 @@ func (suite *KeeperTestSuite) TestMsgCreateDenom() {
...
+		subdenom string
+		valid    bool
+	}{
+		{
+			desc:     "subdenom too long",
+			malleate: func() {},
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/createdenom_test.go:69]: Lets delete the function arg here, and just do it in the main body. (e.g. if `tc.setup != nil { tc.setup() }`)

File: x/tokenfactory/keeper/admins_test.go [hunk 6]
```diff
@@ -1,74 +1,270 @@
...
+func (suite *KeeperTestSuite) TestChangeAdminDenom() {
+	var (
+		denom    string
+		addr0bal int64
+		addr1bal int64
+	)
+
+	for _, tc := range []struct {
+		desc     string
+		malleate func()
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:200]:
```suggestion
		setup func()
```

File: x/tokenfactory/keeper/admins_test.go [hunk 7]
```diff
@@ -1,74 +1,270 @@
...
+				suite.Require().True(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64() == addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
+			} else {
+				suite.Require().Error(err)
+				suite.Require().True(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64() == addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
+			}
+		})
+	}
+}
+
+func (suite *KeeperTestSuite) TestChangeAdminDenom() {
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:191]: This tests logic due to the complex malleate's isn't very easy to follow. I'll spend some more time tmrw thinking about how to restructure it

File: x/tokenfactory/keeper/admins_test.go [hunk 8]
```diff
@@ -1,74 +1,270 @@
...
+
+	// Create a denom and mint
+	suite.CreateDefaultDenom()
+
+	// Make sure that the admin is set correctly
+	queryRes, err := suite.queryClient.DenomAuthorityMetadata(suite.Ctx.Context(), &types.QueryDenomAuthorityMetadataRequest{
+		Denom: suite.defaultDenom,
+	})
+	suite.Require().NoError(err)
+	suite.Require().Equal(suite.TestAccs[0].String(), queryRes.AuthorityMetadata.Admin)
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:133]: Can this be deleted?

File: x/tokenfactory/keeper/admins_test.go [hunk 9]
```diff
@@ -1,74 +1,270 @@
...
+
+	_, err = suite.msgServer.Mint(sdk.WrapSDKContext(suite.Ctx), types.NewMsgMint(suite.TestAccs[0].String(), sdk.NewInt64Coin(suite.defaultDenom, 10)))
+	addr0bal += 10
+	suite.Require().NoError(err)
+	suite.Require().True(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom).Amount.Int64() == addr0bal, suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], suite.defaultDenom))
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:138]: What is this for? Making some tokens to burn?

These lines are hard to follow / the last seems non-standard equality checking syntax)

**Comment by ValarDragon** [comment]
Thank you for making this PR! Apologies for how long it took to get another response, with the chain downtime and restart.

Left some comments on it!

File: x/tokenfactory/keeper/admins_test.go (hunk 8 above)
Comment (vuong177) [reply x/tokenfactory/keeper/admins_test.go:133]: Sure. I deleted it.

**Comment by vuong177** [comment]
@ValarDragon, thanks for your reviews. I've changed code for a better look.

File: x/tokenfactory/keeper/admins_test.go [hunk 10]
```diff
@@ -1,74 +1,262 @@
...
+		},
+	} {
+		suite.Run(fmt.Sprintf("Case %s", tc.desc), func() {
+			// setup test
+			suite.SetupTest()
+			addr0bal = 0
+			addr1bal = 0
+
+			// Create a denom and mint
+			res, err := suite.msgServer.CreateDenom(sdk.WrapSDKContext(suite.Ctx), types.NewMsgCreateDenom(suite.TestAccs[0].String(), "bitcoin"))
```
Comment (mattverse) [inline x/tokenfactory/keeper/admins_test.go:250]: same as comment left in `keeper_test.go` we can either utilzie that better or we can replace that by directly calling the methods from the test, i would prefer the latter option

File: x/tokenfactory/keeper/admins_test.go [hunk 11]
```diff
@@ -1,74 +1,262 @@
...
 	suite.Require().NoError(err)
 	queryRes, err = suite.queryClient.DenomAuthorityMetadata(suite.Ctx.Context(), &types.QueryDenomAuthorityMetadataRequest{
-		Denom: res.GetNewTokenDenom(),
+		Denom: suite.defaultDenom,
 	})
 	suite.Require().NoError(err)
 	suite.Require().Equal("", queryRes.AuthorityMetadata.Admin)
 }
+
+func (suite *KeeperTestSuite) TestMintDenom() {
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:69]:
```suggestion
// TestMintDenom ensures the following properties of the MintMessage:
// * Noone can mint tokens for a denom that doesn't exist
// * Only the admin of a denom can mint tokens for it
// * The admin of a denom can mint tokens for it
func (suite *KeeperTestSuite) TestMintDenom() {
```

File: x/tokenfactory/keeper/admins_test.go (hunk 7 above)
Comment (ValarDragon) [reply x/tokenfactory/keeper/admins_test.go:187]: What if instead of setup(), we added three fields:
* `ChangeAdminMsg`
* `exoectedAdmin`
* `*MintMsg` (and can be nil)
* `expMintPass`

Sorry for the delay it took me to come back to this

**Review by ValarDragon — APPROVED** [review]
Thanks for the PR! I suggest we merge it, and then do the followup on ChangeAdmin tests in a second PR

**Comment by vuong177** [comment]
Yes sir. I'll make a second PR including your suggestion.

**Comment by ValarDragon** [comment]
Thanks so much!
